File: android/app/src/main/java/com/example/shilv/service/TripStore.kt
package com.example.shilv.service

import com.example.shilv.data.DiscoveredTrip
import com.example.shilv.data.MemoryEvent
import com.example.shilv.data.ScanSnapshot
import com.example.shilv.data.TripDates
import com.example.shilv.data.TravelDay
import kotlinx.serialization.json.Json
import java.io.File

/**
 * 旅行索引存储：把 ScanSnapshot 写入本机文件，并保留用户编辑（命名/确认/笔记/精选等）。
 * 逻辑与 iOS TripStore 对齐：重扫时通过 ID 或照片引用匹配旧旅行，合并用户修改。
 */
class TripStore(
    private val storageDirectory: File,
    private val json: Json = Json { ignoreUnknownKeys = true },
) {
    private val file: File = File(storageDirectory, "travel-index-v1.json")

    var snapshot: ScanSnapshot?
        private set

    init {
        storageDirectory.mkdirs()
        load()
    }

    val storedDataSize: Long get() = if (file.exists()) file.length() else 0L

    fun replace(newSnapshot: ScanSnapshot): Result<Unit> = runCatching {
        val merged = mergeUserEdits(old = snapshot, new = newSnapshot)
        persist(merged)
    }

    fun updateTrip(trip: DiscoveredTrip): Result<Unit> = runCatching {
        val value = snapshot ?: return@runCatching
        val reconciled = trip.also { it.reconcileDerivedFields() }
        val index = value.trips.indexOfFirst { it.id == trip.id }
        if (index >= 0) {
            value.trips = value.trips.toMutableList().also { it[index] = reconciled }
        }
        persist(value)
    }

    fun deleteLocalIndex(): Result<Unit> = runCatching {
        if (file.exists()) file.delete()
        snapshot = null
    }

    private fun load() {
        if (!file.exists()) return
        snapshot = runCatching {
            json.decodeFromString<ScanSnapshot>(file.readText())
        }.getOrNull()
    }

    private fun persist(value: ScanSnapshot) {
        file.writeText(json.encodeToString(ScanSnapshot.serializer(), value))
        snapshot = value
    }

    private fun mergeUserEdits(old: ScanSnapshot?, new: ScanSnapshot): ScanSnapshot {
        if (old == null) return new
        val previousTrips = old.trips.associateBy { it.id }
        val merged = new.copy()

        merged.trips = new.trips.map { fresh ->
            val previous = previousTrips[fresh.id]
                ?: old.trips.filter { matchScore(it, fresh) >= 0.5 }.maxByOrNull { matchScore(it, fresh) }
            if (previous == null) return@map fresh
            mergeTrip(previous, fresh)
        }

        val matchedOldIDs = merged.trips.mapNotNull { fresh ->
            old.trips.filter { matchScore(it, fresh) >= 0.5 }.maxByOrNull { matchScore(it, fresh) }?.id
        }.toSet()
        val preservedMemories = old.trips.filter { it.isConfirmed && !matchedOldIDs.contains(it.id) }
        merged.trips = (merged.trips + preservedMemories).sortedByDescending { it.startDate }
        return merged
    }

    private fun mergeTrip(previous: DiscoveredTrip, fresh: DiscoveredTrip): DiscoveredTrip {
        val result = fresh.copy()
        result.title = previous.title
        result.isConfirmed = previous.isConfirmed
        result.isHidden = previous.isHidden
        result.summary = previous.summary
        result.isFavorite = previous.isFavorite
        result.coverPhotoIDOverride = previous.coverPhotoIDOverride
        result.suppressedEventIDs = previous.suppressedEventIDs

        val oldEvents = previous.days.flatMap { it.events }.associateBy { it.id }
        result.days = fresh.days.map { day ->
            var updated = day.copy()
            val oldDay = previous.days.minByOrNull { kotlin.math.abs(it.date - day.date) }
            if (oldDay != null && kotlin.math.abs(oldDay.date - day.date) < 12 * 60 * 60 * 1000L) {
                updated = updated.copy(title = oldDay.title)
            }
            updated = updated.copy(events = day.events.map { event ->
                val oldEvent = oldEvents[event.id]
                    ?: previous.days.flatMap { it.events }.filter { kotlin.math.abs(it.startDate - event.startDate) <= 30 * 60 * 1000L }
                        .minByOrNull { kotlin.math.abs(it.startDate - event.startDate) }
                if (oldEvent == null) return@map event
                mergeEvent(oldEvent, event)
            })
            val customEvents = previous.days.flatMap { it.events }
                .filter { it.isUserCreated == true && TripDates.startOfDay(it.startDate) == TripDates.startOfDay(day.date) }
            val existingIDs = updated.events.map { it.id }.toSet()
            updated = updated.copy(events = updated.events + customEvents.filter { it.id !in existingIDs })
            removeSuppressed(updated, previous.suppressedEventIDs)
        }

        val existingEventIDs = result.days.flatMap { it.events }.map { it.id }.toSet()
        val unmatchedCustomEvents = previous.days.flatMap { it.events }
            .filter { it.isUserCreated == true && it.id !in existingEventIDs }
        for (custom in unmatchedCustomEvents) {
            val dayIndex = result.days.indexOfFirst { TripDates.startOfDay(it.date) == TripDates.startOfDay(custom.startDate) }
            if (dayIndex >= 0) {
                result.days = result.days.toMutableList().also {
                    it[dayIndex] = it[dayIndex].copy(events = it[dayIndex].events + custom)
                }
            } else {
                result.days = result.days + TravelDay(
                    id = custom.id + "-day", date = TripDates.startOfDay(custom.startDate),
                    title = "补充的一天", events = listOf(custom),
                )
            }
        }
        result.days = result.days.map { removeSuppressed(it, previous.suppressedEventIDs) }
        result.days = normalizeEventDays(result.days)
        result.reconcileDerivedFields()
        return result
    }

    private fun mergeEvent(old: MemoryEvent, fresh: MemoryEvent): MemoryEvent {
        val value = fresh.copy()
        value.title = old.title
        value.placeName = old.placeName
        value.cityName = old.cityName
        value.countryName = old.countryName
        value.note = old.note
        value.isHidden = old.isHidden
        value.startDate = old.startDate
        value.endDate = old.endDate
        value.location = old.location
        value.summary = old.summary
        value.featuredPhotoIDs = old.featuredPhotoIDs
        value.coverPhotoIDOverride = old.coverPhotoIDOverride
        value.excludedPhotoIDs = old.excludedPhotoIDs
        value.isUserCreated = old.isUserCreated
        val existing = value.photoIDs
        value.photoIDs = value.photoIDs + old.photoIDs.filter { it !in existing }
        old.excludedPhotoIDs?.let { excluded ->
            value.photoIDs = value.photoIDs.filter { it !in excluded }
            value.featuredPhotoIDs = value.featuredPhotoIDs?.filter { it !in excluded }
        }
        return value
    }

    private fun removeSuppressed(day: TravelDay, suppressed: List<String>?): TravelDay =
        if (suppressed == null) day else day.copy(events = day.events.filter { it.id !in suppressed })

    private fun normalizeEventDays(days: List<TravelDay>): List<TravelDay> {
        val grouped = days.flatMap { it.events }.groupBy { TripDates.startOfDay(it.startDate) }
        return grouped.keys.sorted().map { date ->
            val existing = days.firstOrNull { TripDates.startOfDay(it.date) == date }
            TravelDay(
                id = existing?.id ?: "day-$date",
                date = date,
                title = existing?.title ?: "补充的一天",
                events = grouped[date].orEmpty().sortedBy { it.startDate },
            )
        }
    }

    private fun overlapRatio(a: DiscoveredTrip, b: DiscoveredTrip): Double {
        val intersectionStart = maxOf(a.startDate, b.startDate)
        val intersectionEnd = minOf(a.endDate, b.endDate)
        if (intersectionEnd <= intersectionStart) return 0.0
        val intersection = intersectionEnd - intersectionStart
        val smaller = minOf(a.endDate - a.startDate, b.endDate - b.startDate)
        return if (smaller > 0) intersection.toDouble() / smaller else 0.0
    }

    private fun photoOverlapRatio(a: DiscoveredTrip, b: DiscoveredTrip): Double {
        val left = a.photoIDs.toSet()
        val right = b.photoIDs.toSet()
        if (left.isEmpty() || right.isEmpty()) return 0.0
        return left.intersect(right).size.toDouble() / minOf(left.size, right.size)
    }

    private fun matchScore(a: DiscoveredTrip, b: DiscoveredTrip): Double {
        if (a.id == b.id) return 1.0
        return maxOf(overlapRatio(a, b), photoOverlapRatio(a, b))
    }
}

File: ios/ShiLv/Services/TripStore.swift
import Foundation
import Combine

@MainActor
final class TripStore: ObservableObject {
    @Published private(set) var snapshot: ScanSnapshot?
    private let fileURL: URL

    init(fileManager: FileManager = .default, storageDirectory: URL? = nil) {
        let root = storageDirectory
            ?? fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? fileManager.temporaryDirectory
        let directory = storageDirectory == nil ? root.appending(path: "ShiLv", directoryHint: .isDirectory) : root
        try? fileManager.createDirectory(at: directory, withIntermediateDirectories: true)
        fileURL = directory.appending(path: "travel-index-v1.json")
        load()
    }

    func replace(with newSnapshot: ScanSnapshot) throws {
        let merged = mergeUserEdits(old: snapshot, new: newSnapshot)
        try persist(merged)
    }

    func updateTrip(_ trip: DiscoveredTrip) throws {
        guard var value = snapshot, let index = value.trips.firstIndex(where: { $0.id == trip.id }) else { return }
        var reconciled = trip
        reconciled.reconcileDerivedFields()
        value.trips[index] = reconciled
        try persist(value)
    }

    func deleteLocalIndex() throws {
        if FileManager.default.fileExists(atPath: fileURL.path) { try FileManager.default.removeItem(at: fileURL) }
        snapshot = nil
    }

    var storedDataSize: Int64 {
        Int64((try? fileURL.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0)
    }

    private func load() {
        guard let data = try? Data(contentsOf: fileURL), let decoded = try? JSONDecoder.shilv.decode(ScanSnapshot.self, from: data) else { return }
        snapshot = decoded
    }

    private func persist(_ value: ScanSnapshot) throws {
        let data = try JSONEncoder.shilv.encode(value)
        try data.write(to: fileURL, options: [.atomic, .completeFileProtection])
        snapshot = value
    }

    private func mergeUserEdits(old: ScanSnapshot?, new: ScanSnapshot) -> ScanSnapshot {
        guard let old else { return new }
        let previousTrips = Dictionary(uniqueKeysWithValues: old.trips.map { ($0.id, $0) })
        var merged = new
        merged.trips = new.trips.map { fresh in
            let previous = previousTrips[fresh.id] ?? old.trips.max(by: {
                matchScore($0, fresh) < matchScore($1, fresh)
            }).flatMap { matchScore($0, fresh) >= 0.5 ? $0 : nil }
            guard let previous else { return fresh }
            var result = fresh
            result.title = previous.title
            result.isConfirmed = previous.isConfirmed
            result.isHidden = previous.isHidden
            result.summary = previous.summary
            result.isFavorite = previous.isFavorite
            result.coverPhotoIDOverride = previous.coverPhotoIDOverride
            result.suppressedEventIDs = previous.suppressedEventIDs
            let oldEvents = Dictionary(uniqueKeysWithValues: previous.days.flatMap(\.events).map { ($0.id, $0) })
            result.days = fresh.days.map { day in
                var updated = day
                if let oldDay = previous.days.min(by: {
                    abs($0.date.timeIntervalSince(day.date)) < abs($1.date.timeIntervalSince(day.date))
                }), abs(oldDay.date.timeIntervalSince(day.date)) < 12 * 60 * 60 {
                    updated.title = oldDay.title
                }
                updated.events = day.events.map { event in
                    let oldEvent = oldEvents[event.id] ?? previous.days.flatMap(\.events).min(by: {
                        abs($0.startDate.timeIntervalSince(event.startDate)) < abs($1.startDate.timeIntervalSince(event.startDate))
                    }).flatMap { abs($0.startDate.timeIntervalSince(event.startDate)) <= 30 * 60 ? $0 : nil }
                    guard let oldEvent else { return event }
                    var value = event
                    value.title = oldEvent.title; value.placeName = oldEvent.placeName
                    value.cityName = oldEvent.cityName; value.countryName = oldEvent.countryName
                    value.note = oldEvent.note; value.isHidden = oldEvent.isHidden
                    value.startDate = oldEvent.startDate; value.endDate = oldEvent.endDate
                    value.location = oldEvent.location
                    value.summary = oldEvent.summary
                    value.featuredPhotoIDs = oldEvent.featuredPhotoIDs
                    value.coverPhotoIDOverride = oldEvent.coverPhotoIDOverride
                    value.excludedPhotoIDs = oldEvent.excludedPhotoIDs
                    value.isUserCreated = oldEvent.isUserCreated
                    let existingPhotoIDs = value.photoIDs
                    value.photoIDs += oldEvent.photoIDs.filter { !existingPhotoIDs.contains($0) }
                    if let excluded = oldEvent.excludedPhotoIDs {
                        value.photoIDs.removeAll { excluded.contains($0) }
                        value.featuredPhotoIDs?.removeAll { excluded.contains($0) }
                    }
                    return value
                }
                let customEvents = previous.days.flatMap(\.events).filter { $0.isUserCreated == true && $0.startDate.startOfDay == day.date.startOfDay }
                let existingIDs = Set(updated.events.map(\.id))
                updated.events += customEvents.filter { !existingIDs.contains($0.id) }
                if let suppressed = previous.suppressedEventIDs { updated.events.removeAll { suppressed.contains($0.id) } }
                return updated
            }
            let existingEventIDs = Set(result.days.flatMap(\.events).map(\.id))
            let unmatchedCustomEvents = previous.days.flatMap(\.events).filter { $0.isUserCreated == true && !existingEventIDs.contains($0.id) }
            for custom in unmatchedCustomEvents {
                if let dayIndex = result.days.firstIndex(where: { $0.date.startOfDay == custom.startDate.startOfDay }) {
                    result.days[dayIndex].events.append(custom)
                } else {
                    result.days.append(TravelDay(id: UUID(), date: custom.startDate.startOfDay, title: "补充的一天", events: [custom]))
                }
            }
            if let suppressed = previous.suppressedEventIDs {
                for index in result.days.indices { result.days[index].events.removeAll { suppressed.contains($0.id) } }
            }
            result.days = normalizeEventDays(result.days)
            result.reconcileDerivedFields()
            return result
        }
        let matchedOldIDs = Set(merged.trips.compactMap { fresh in
            old.trips.max(by: { matchScore($0, fresh) < matchScore($1, fresh) }).flatMap { matchScore($0, fresh) >= 0.5 ? $0.id : nil }
        })
        let preservedMemories = old.trips.filter { $0.isConfirmed && !matchedOldIDs.contains($0.id) }
        merged.trips += preservedMemories
        merged.trips.sort { $0.startDate > $1.startDate }
        return merged
    }

    private func normalizeEventDays(_ days: [TravelDay]) -> [TravelDay] {
        let allEvents = days.flatMap(\.events)
        let grouped = Dictionary(grouping: allEvents) { $0.startDate.startOfDay }
        return grouped.keys.sorted().map { date in
            let existing = days.first(where: { $0.date.startOfDay == date })
            return TravelDay(
                id: existing?.id ?? UUID(),
                date: date,
                title: existing?.title ?? "补充的一天",
                events: (grouped[date] ?? []).sorted { $0.startDate < $1.startDate }
            )
        }
    }

    private func overlapRatio(_ lhs: DiscoveredTrip, _ rhs: DiscoveredTrip) -> Double {
        let intersectionStart = max(lhs.startDate, rhs.startDate)
        let intersectionEnd = min(lhs.endDate, rhs.endDate)
        guard intersectionEnd > intersectionStart else { return 0 }
        let intersection = intersectionEnd.timeIntervalSince(intersectionStart)
        let smallerDuration = min(lhs.endDate.timeIntervalSince(lhs.startDate), rhs.endDate.timeIntervalSince(rhs.startDate))
        return smallerDuration > 0 ? intersection / smallerDuration : 0
    }

    private func photoOverlapRatio(_ lhs: DiscoveredTrip, _ rhs: DiscoveredTrip) -> Double {
        let left = Set(lhs.photoIDs), right = Set(rhs.photoIDs)
        guard !left.isEmpty, !right.isEmpty else { return 0 }
        return Double(left.intersection(right).count) / Double(min(left.count, right.count))
    }

    private func matchScore(_ lhs: DiscoveredTrip, _ rhs: DiscoveredTrip) -> Double {
        if lhs.id == rhs.id { return 1 }
        return max(overlapRatio(lhs, rhs), photoOverlapRatio(lhs, rhs))
    }
}

private extension JSONEncoder {
    static var shilv: JSONEncoder { let encoder = JSONEncoder(); encoder.dateEncodingStrategy = .iso8601; encoder.outputFormatting = [.sortedKeys]; return encoder }
}
private extension JSONDecoder {
    static var shilv: JSONDecoder { let decoder = JSONDecoder(); decoder.dateDecodingStrategy = .iso8601; return decoder }
}
